src/query/core/buildMiddleware/cacheLifecycle.ts

```diff
@@ -0,0 +1,241 @@
+
+declare module '../../endpointDefinitions' {
+  export type LifecycleApi<
+    ReducerPath extends string = string,
+    CacheEntry = unknown
+  > = {
+    /**
+     * The dispatch method for the store
+     */
+    dispatch: ThunkDispatch<any, any, AnyAction>
+    /**
+     * A method to get the current state
+     */
+    getState(): RootState<any, any, ReducerPath>
+    /**
+     * `extra` as provided as `thunk.extraArgument` to the `configureStore` `getDefaultMiddleware` option.
+     */
+    extra: unknown
+    /**
+     * A unique ID generated for the mutation
+     */
+    requestId: string
+    /**
+     * Gets the current value of this cache entry.
+     */
+    getCacheEntry: () => CacheEntry
+  }
+
+  export interface CacheLifecyclePromises<ResultType = unknown> {
+    /**
+     * Promise that will resolve with the first value for this cache key.
+     * This allows you to `await` until an actual value is in cache.
+     *
+     * If the cache entry is removed from the cache before any value has ever
+     * been resolved, this Promise will reject with
+     * `new Error('Promise never resolved before cleanup.')`
+     * to prevent memory leaks.
+     * You can just re-throw that error (or not handle it at all) -
+     * it will be caught outside of `cacheEntryAdded`.
+     */
+    firstValueResolved: OptionalPromise<ResultType>
+    /**
+     * Promise that allows you to wait for the point in time when the cache entry
+     * has been removed from the cache, by not being used/subscribed to any more
+     * in the application for too long or by dispatching `api.util.resetApiState`.
+     */
+    cleanup: Promise<void>
+  }
+
+  interface QueryExtraOptions<
+    TagTypes extends string,
+    ResultType,
+    QueryArg,
+    BaseQuery extends BaseQueryFn,
+    ReducerPath extends string = string
+  > {
+    onCacheEntryAdded?(
+      arg: QueryArg,
+      api: LifecycleApi<
+        ReducerPath,
+        QueryResultSelectorResult<
+          { type: DefinitionType.query } & BaseEndpointDefinition<
+            QueryArg,
+            BaseQuery,
+            ResultType
+          >
+        >
+      >,
+      promises: CacheLifecyclePromises<ResultType>
+    ): Promise<void> | void
+  }
+
+  interface MutationExtraOptions<
+    TagTypes extends string,
+    ResultType,
+    QueryArg,
+    BaseQuery extends BaseQueryFn,
+    ReducerPath extends string = string
+  > {
+    onCacheEntryAdded?(
+      arg: QueryArg,
+      api: LifecycleApi<
+        ReducerPath,
+        MutationResultSelectorResult<
+          { type: DefinitionType.mutation } & BaseEndpointDefinition<
+            QueryArg,
+            BaseQuery,
+            ResultType
+          >
+        >
+      >,
+      promises: CacheLifecyclePromises<ResultType>
+    ): Promise<void> | void
+  }
+}
+
+export const build: SubMiddlewareBuilder = ({
+  api,
+  reducerPath,
+  context,
+  queryThunk,
+  mutationThunk,
+}) => {
+  type CacheLifecycle = {
```

Comment:
Can this be moved outside, or is there a reason it's declared here?

Reply:
It's only used in a locally scoped variable in the next line and once more to declare a value to add to that variable.

Moving it outside would raise questions like "should this be exported" etc, just keeping it here just circumvents all that and will also 100% remove it on transpilation.

Reply:
Cool. Tx for the explanation.